### ISSUE TITLE
Binary cross entropy and CuArrays errors

### ISSUE THREAD

**Issue by maetshju** [FluxML/Flux.jl#275]
I've experienced two different errors when using binary cross entropy loss in conjunction with CuArrays for GPU processing. I'm unsure if they are completely related or not, but I figured I could move one of them to a separate issue if needed. Here is a short script to reproduce them:

```julia
using Flux
using Flux: binarycrossentropy, relu, sigmoid
using CuArrays

Xs = Vector()
Ys = Vector()

for i=1:5
	x = rand(500, 20)
	x = [x[i,:] for i in size(x,1)]
	x = gpu.(x)
	push!(Xs, x)
	push!(Ys, gpu(bitrand(500)))
end

data = collect(zip(Xs, Ys))

m = Chain(Dense(20, 64, relu),
	   Dense(64, 1, sigmoid)) |> gpu 

function loss(x, y)
	println("predicting")
	yhats = m.(x)
	println("calculating bce")
	l = sum(binarycrossentropy.(yhats, y))
	println("done calculating bce")
	#=bces = Vector()
	for (yhati, yi) in zip(yhats, y)
		println("calculating bce")
		push!(bces, binarycrossentropy(yhati, yi))
		println("done calculating bce")
	end=#
	l = sum(bces)
	return l
end

opt = ADAM(params(m))

Flux.train!(loss, data, opt)
```

When using the section of `loss` that uses dot broadcasting, I get the following error:

```
ERROR: LoadError: don't know how to handle argument of type Array{TrackedArray{…,CuArray{Float32,1}},0}
Stacktrace:
 [1] cudaconvert(::Array{TrackedArray{…,CuArray{Float32,1}},0}) at /home/maetshju/.julia/v0.6/CUDAnative/src/execution.jl:20
 [2] map at ./tuple.jl:159 [inlined]
 [3] broadcast at ./broadcast.jl:17 [inlined]
 [4] cudaconvert at /home/maetshju/.julia/v0.6/CUDAnative/src/execution.jl:26 [inlined]
 [5] map(::CUDAnative.#cudaconvert, ::Tuple{Tuple{UInt32},Tuple{GPUArrays.BInfo{Array,0},GPUArrays.BInfo{CuArrays.CuArray,1}},Tuple{Array{TrackedArray{…,CuArray{Float32,1}},0},CuArray{Float32,1}}}) at ./tuple.jl:160
 [6] map at ./tuple.jl:161 [inlined] (repeats 4 times)
 [7] broadcast at ./broadcast.jl:17 [inlined]
 [8] _gpu_call(::Function, ::CuArray{TrackedArray{…,CuArray{Float32,1}},1}, ::Tuple{Flux.#binarycrossentropy,CuArray{TrackedArray{…,CuArray{Float32,1}},1},Tuple{UInt32},Tuple{GPUArrays.BInfo{Array,0},GPUArrays.BInfo{CuArrays.CuArray,1}},Tuple{Array{TrackedArray{…,CuArray{Float32,1}},0},CuArray{Float32,1}}}, ::Tuple{Tuple{Int64},Tuple{Int64}}) at /home/maetshju/.julia/v0.6/CuArrays/src/gpuarray_interface.jl:61
 [9] gpu_call(::Function, ::CuArray{TrackedArray{…,CuArray{Float32,1}},1}, ::Tuple{Flux.#binarycrossentropy,CuArray{TrackedArray{…,CuArray{Float32,1}},1},Tuple{UInt32},Tuple{GPUArrays.BInfo{Array,0},GPUArrays.BInfo{CuArrays.CuArray,1}},Tuple{Array{TrackedArray{…,CuArray{Float32,1}},0},CuArray{Float32,1}}}, ::Int64) at /home/maetshju/.julia/v0.6/GPUArrays/src/abstract_gpu_interface.jl:151
 [10] _broadcast!(::Function, ::CuArray{TrackedArray{…,CuArray{Float32,1}},1}, ::Tuple{Tuple{},Tuple{Bool}}, ::Tuple{Tuple{},Tuple{Int64}}, ::Array{TrackedArray{…,CuArray{Float32,1}},0}, ::Tuple{CuArray{Float32,1}}, ::Type{Val{0}}, ::CartesianRange{CartesianIndex{1}}) at /home/maetshju/.julia/v0.6/GPUArrays/src/broadcast.jl:89
 [11] broadcast_t(::Function, ::Type{T} where T, ::Tuple{Base.OneTo{Int64}}, ::CartesianRange{CartesianIndex{1}}, ::Array{TrackedArray{…,CuArray{Float32,1}},0}, ::CuArray{Float32,1}) at /home/maetshju/.julia/v0.6/GPUArrays/src/broadcast.jl:68
 [12] broadcast_c(::Function, ::Type{Array}, ::Array{TrackedArray{…,CuArray{Float32,1}},0}, ::CuArray{Float32,1}, ::Vararg{CuArray{Float32,1},N} where N) at ./broadcast.jl:316
 [13] loss(::Array{CuArray{Float32,1},0}, ::CuArray{Float32,1}) at /home/maetshju/gsoc2018/speech-cnn/testbince.jl:25
 [14] macro expansion at /home/maetshju/.julia/v0.6/Flux/src/optimise/train.jl:39 [inlined]
 [15] macro expansion at /home/maetshju/.julia/v0.6/Juno/src/progress.jl:119 [inlined]
 [16] #train!#130(::Flux.Optimise.##131#133, ::Function, ::Function, ::Array{Tuple{Any,Any},1}, ::Flux.Optimise.##71#75) at /home/maetshju/.julia/v0.6/Flux/src/optimise/train.jl:38
 [17] train!(::Function, ::Array{Tuple{Any,Any},1}, ::Function) at /home/maetshju/.julia/v0.6/Flux/src/optimise/train.jl:36
 [18] include_from_node1(::String) at ./loading.jl:576
 [19] include(::String) at ./sysimg.jl:14
 [20] process_options(::Base.JLOptions) at ./client.jl:305
 [21] _start() at ./client.jl:371
while loading /home/maetshju/gsoc2018/speech-cnn/testbince.jl, in expression starting on line 39
```

And when using the section with the for-loop for calculating binary cross entropy, I get this error output:

```
ERROR: LoadError: CUDA error: an illegal memory access was encountered (code #700, ERROR_ILLEGAL_ADDRESS)
Stacktrace:
 [1] macro expansion at /home/maetshju/.julia/v0.6/CUDAdrv/src/base.jl:148 [inlined]
 [2] CUDAdrv.CuModule(::String, ::Dict{CUDAdrv.CUjit_option,Any}) at /home/maetshju/.julia/v0.6/CUDAdrv/src/module.jl:35
 [3] cufunction(::CUDAdrv.CuDevice, ::Any, ::Any) at /home/maetshju/.julia/v0.6/CUDAnative/src/jit.jl:500
 [4] macro expansion at /home/maetshju/.julia/v0.6/CUDAnative/src/execution.jl:107 [inlined]
 [5] _cuda(::Tuple{Int64,Int64}, ::Int64, ::CUDAdrv.CuStream, ::CuArrays.#broadcast_kernel, ::Flux.Tracker.##33#34, ::CUDAnative.CuDeviceArray{Float32,1,CUDAnative.AS.Global}, ::Tuple{Tuple{Bool}}, ::Tuple{Tuple{Int64}}, ::CUDAnative.CuDeviceArray{ForwardDiff.Dual{Void,Float32,1},1,CUDAnative.AS.Global}, ::Tuple{}) at /home/maetshju/.julia/v0.6/CUDAnative/src/execution.jl:80
 [6] _broadcast! at /home/maetshju/.julia/v0.6/CuArrays/src/broadcast.jl:22 [inlined]
 [7] broadcast_t at /home/maetshju/.julia/v0.6/CuArrays/src/broadcast.jl:37 [inlined]
 [8] broadcast_c at /home/maetshju/.julia/v0.6/CuArrays/src/broadcast.jl:58 [inlined]
 [9] broadcast at ./broadcast.jl:455 [inlined]
 [10] map(::Function, ::CuArray{ForwardDiff.Dual{Void,Float32,1},1}) at /home/maetshju/.julia/v0.6/CuArrays/src/utils.jl:62
 [11] (::Flux.Tracker.Broadcasted{Base.#log,CuArray{ForwardDiff.Dual{Void,Float32,1},1}})() at /home/maetshju/.julia/v0.6/Flux/src/tracker/array.jl:270
 [12] tracked_broadcast(::Function, ::TrackedArray{…,CuArray{Float32,1}}) at /home/maetshju/code sample.julia/v0.6/Flux/src/tracker/array.jl:281
 [13] log(::TrackedArray{…,CuArray{Float32,1}}) at ./deprecated.jl:59
 [14] loss(::Array{CuArray{Float32,1},0}, ::CuArray{Float32,1}) at /home/maetshju/gsoc2018/speech-cnn/testbince.jl:30
 [15] macro expansion at /home/maetshju/.julia/v0.6/Flux/src/optimise/train.jl:39 [inlined]
 [16] macro expansion at /home/maetshju/.julia/v0.6/Juno/src/progress.jl:119 [inlined]
 [17] #train!#130(::Flux.Optimise.##131#133, ::Function, ::Function, ::Array{Tuple{Any,Any},1}, ::Flux.Optimise.##71#75) at /home/maetshju/.julia/v0.6/Flux/src/optimise/train.jl:38
 [18] train!(::Function, ::Array{Tuple{Any,Any},1}, ::Function) at /home/maetshju/.julia/v0.6/Flux/src/optimise/train.jl:36
 [19] include_from_node1(::String) at ./loading.jl:576
 [20] include(::String) at ./sysimg.jl:14
 [21] process_options(::Base.JLOptions) at ./client.jl:305
 [22] _start() at ./client.jl:371
while loading /home/maetshju/gsoc2018/speech-cnn/testbince.jl, in expression starting on line 39
error in running finalizer: CUDAdrv.CuError(code=700, meta=nothing)
error in running finalizer: CUDAdrv.CuError(code=700, meta=nothing)
error in running finalizer: CUDAdrv.CuError(code=700, meta=nothing)
```

I'm not sure, but I believe that this second error may be related to https://github.com/JuliaGPU/CuArrays.jl/issues/48. Neither of these issues seemed to crop up for me when I switched to mse loss.

I'm using Flux 0.5.1 and CuArrays 0.5.0, both of which I just installed today.

**Comment by jekbradbury** [comment]
Maybe try again with Flux and CuArrays master (i.e., `Pkg.checkout("Flux"); Pkg.checkout("CuArrays")`? I think some issues similar to the first one you saw have been solved recently. In general though errors like that are pretty hard to track down (usually method ambiguity issues related to Tracker), so you'll probably have to get Mike's attention at some point if it's not fixed on master.

**Comment by maetshju** [comment]
Thanks for the suggestion, @jekbradbury! I tried checking out the master versions of each package individually, and then at the same time, and the same errors occurred in each scenario, unfortunately (the table below describes this if I wrote that confusingly).

— | Flux release | Flux master
 ----| ------------------|-------------------
**CuArrays release** | Error | Error
**CuArrays master** | Error | Error

I was hoping that if one of these checkouts did fix the issue, we might be able to narrow down which package it came from, but not such luck.
